presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -760,16 +756,13 @@ public PlanWithProperties visitJoin(JoinNode node, Context context)
             PlanWithProperties left;
             PlanWithProperties right;
 
-            boolean isCrossJoin = type == INNER && leftSymbols.isEmpty();
-            if ((distributedJoins && !isCrossJoin && !isScalar(node.getRight())) || (type == FULL) || (type == RIGHT)) {
-                // The implementation of full outer join only works if the data is hash partitioned. See LookupJoinOperators#buildSideOuterJoinUnvisitedPositions
-
+            if (node.isDistributed()) {
                 SetMultimap<Symbol, Symbol> rightToLeft = createMapping(rightSymbols, leftSymbols);
                 SetMultimap<Symbol, Symbol> leftToRight = createMapping(leftSymbols, rightSymbols);
 
                 left = node.getLeft().accept(this, context.withPreferredProperties(PreferredProperties.partitioned(ImmutableSet.copyOf(leftSymbols))));
 
-                if (left.getProperties().isNodePartitionedOn(leftSymbols) && !(left.getProperties().isSingleNode() && distributedJoins)) {
+                if (left.getProperties().isNodePartitionedOn(leftSymbols) && !left.getProperties().isSingleNode()) {
```

Comment:
I assumed this one was also intended to be the same as the condition in its enclosing if-statement (and therefore also redundant).  It doesn't make sense to have different behavior for full/right outer joins when distributed_joins is set to false vs. true, since those join types are always distributed.